gdal/ogr/ogrsf_frmts/cad/libopencad/cadgeometry.h

```diff
@@ -0,0 +1,666 @@
+{
+public:
+    CADUnknown();
+
+    virtual void print() const override;
+    void         transform( const Matrix& matrix ) override;
+};
+
+/**
+ * @brief Geometry class which a single Point
+ */
+class CADPoint3D : public CADGeometry
+{
+public:
+              CADPoint3D();
+              CADPoint3D( const CADVector& positionIn, double thicknessIn );
+    CADVector getPosition() const;
+    void      setPosition( const CADVector& value );
+
+    CADVector getExtrusion() const;
+    void      setExtrusion( const CADVector& value );
+
+    double getXAxisAng() const;
+    void   setXAxisAng( double value );
+
+    virtual void print() const override;
+    virtual void transform( const Matrix& matrix ) override;
+protected:
+    CADVector position;
+    CADVector extrusion;
+    double    xAxisAng;
+};
+
+/**
+ * @brief Geometry class which represents a simple Line
+ */
+class CADLine : public CADGeometry
+{
+public:
+    CADLine();
+    CADLine( const CADPoint3D& startIn, const CADPoint3D& endIn );
+
+    CADPoint3D getStart() const;
+    void       setStart( const CADPoint3D& value );
+
+    CADPoint3D getEnd() const;
+    void       setEnd( const CADPoint3D& value );
+
+    virtual void print() const override;
+    virtual void transform( const Matrix& matrix ) override;
+protected:
+    CADPoint3D start;
+    CADPoint3D end;
+};
+
+/**
+ * @brief Geometry class which represents Polyline 3D
+ */
+class CADPolyline3D : public CADGeometry
+{
+public:
+    CADPolyline3D();
+
+    void   addVertex( const CADVector& vertex );
+    size_t getVertexCount() const;
+    CADVector& getVertex( size_t index );
+
+    virtual void print() const override;
+    virtual void transform( const Matrix& matrix ) override;
+protected:
+    vector<CADVector> vertexes;
+};
+
+/**
+ * @brief Geometry class which represents LWPolyline
+ */
+
+class CADLWPolyline : public CADPolyline3D
+{
+public:
+    CADLWPolyline();
+
+    double getConstWidth() const;
+    void   setConstWidth( double value );
+
+    double getElevation() const;
+    void   setElevation( double value );
+
+    CADVector getVectExtrusion() const;
+    void      setVectExtrusion( const CADVector& value );
+
+    vector<pair<double, double> > getWidths() const;
+    void                          setWidths( const vector<pair<double, double> >& value );
+
+    vector<double> getBulges() const;
+    void           setBulges( const vector<double>& value );
+
+    bool isClosed() const;
+    void setClosed( bool state );
+
+    virtual void print() const override;
+protected:
+    bool                          bClosed;
+    double                        constWidth;
+    double                        elevation;
+    CADVector                     vectExtrusion;
+    vector<double>                bulges;
+    vector<pair<double, double> > widths; // start, end.
+};
+
+/**
+ * @brief Geometry class which represents Circle
+ */
+class CADCircle : public CADPoint3D
+{
+public:
+    CADCircle();
```

Comment:
Could you simplify code elsewhere and possibly have const objects if your geometries had constructors that to args?